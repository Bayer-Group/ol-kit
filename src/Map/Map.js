import React from 'react'
import PropTypes from 'prop-types'
import nanoid from 'nanoid'
import debounce from 'lodash.debounce'

import MapLogo from './MapLogo'
import {
  createMap,
  updateMapFromUrl,
  updateUrlFromMap,
  replaceZoomBoxCSS,
  createSelectInteraction
} from './utils'
import { StyledMap } from './styled'
import { connectToContext } from 'Provider'
import en from 'locales/en'
import ugh from 'ugh'

/**
 * A Reactified ol.Map wrapper component
 * @component
 * @category Map
 * @since 0.1.0
 */
class Map extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      mapInitialized: false
    }

    // map is passed as a prop- use this flag to determine whether a map/portal should be created
    this.passedMap = props.map

    // this is used to create a unique identifier for the map div
    this.target = `_ol_kit_map_${nanoid(6)}`
  }

  componentDidMount () {
    const {
      addMapToContext,
      map: passedMap,
      onMapInit,
      translations,
      updateUrlDebounce,
      updateUrlFromView,
      updateViewFromUrl,
      urlViewParam,
      dragZoomboxStyle
    } = this.props
    const onMapReady = map => {
      // pass map back via callback prop
      const initCallback = onMapInit(map)
      // if onMapInit prop returns a promise, render children after promise is resolved
      const isPromise = !!initCallback && typeof initCallback.then === 'function'

      // update AFTER onMapInit to get map into the state/context
      isPromise
        ? initCallback
          .catch(e => ugh.error('Error caught in \'onMapInit\'', e))
          .finally(() => this.setState({ mapInitialized: true })) // always initialize app
        : this.setState({ mapInitialized: true })
    }

    // if no map was passed, create the map
    this.map = !this.passedMap ? createMap({ target: this.target }) : passedMap

    // setup select interactions for the map
    this.initializeSelect(this.map)

    // callback for <Provider> if it is mounted as hoc
    const mapConfig = {
      map: this.map,
      mapId: this.target,
      selectInteraction: this.selectInteraction,
      translations // this can be hoisted to <Provider> only in the future
    }

    addMapToContext(mapConfig)

    // optionally attach map listener
    if (updateUrlFromView) {
      const setUrl = () => updateUrlFromMap(this.map, urlViewParam)
      const mapMoveListener = debounce(setUrl, updateUrlDebounce)

      // update the url param after map movements
      this.map.on('moveend', mapMoveListener)
    }

    // optionally update map view from url param
    if (updateViewFromUrl) {
      // read the url to update the map from view info
      updateMapFromUrl(this.map, urlViewParam)
        .catch(ugh.error)
        .finally(() => onMapReady(this.map)) // always fire callback with map reference on success/failure
    } else {
      // callback that returns a reference to the created map
      onMapReady(this.map)
    }

    // optionally add zoombox styling
    replaceZoomBoxCSS(dragZoomboxStyle)
  }

  initializeSelect = map => {
    const { selectInteraction } = this.props

    if (selectInteraction) {
      // if select is passed as a prop always use that one first
      this.selectInteraction = selectInteraction
    } else {
      // otherwise create a new select interaction
      this.selectInteraction = createSelectInteraction()
    }

    // check the map to see if select interaction has been added
    const selectInteractionOnMap = map.getInteractions().getArray()
      // Layer panel also adds a select interaction
      .filter(interaction => interaction._ol_kit_interaction_type !== '_ol_kit_layer_panel_hover')
      // this checks if the select interaction created or passed in is the same instance on the map and never double adds
      .find(interaction => interaction === this.selectInteraction)

    // do not double add the interaction to the map
    if (!selectInteractionOnMap) map.addInteraction(this.selectInteraction)
  }

  render () {
    const { children, fullScreen, logoPosition, style, translations } = this.props
    const { mapInitialized } = this.state

    return (
      <>
        {!this.passedMap &&
          <StyledMap
            id={this.target}
            fullScreen={fullScreen}
            style={style}>
            <MapLogo logoPosition={logoPosition} translations={translations} />
          </StyledMap>
        }
        {mapInitialized // wait for map to initialize before rendering children
          ? children
          : null
        }
      </>
    )
  }
}

Map.defaultProps = {
  addMapToContext: () => {},
  fullScreen: false,
  logoPosition: 'right',
  map: null,
  onMapInit: () => {},
  updateUrlDebounce: 400,
  updateUrlFromView: true,
  updateViewFromUrl: true,
  urlViewParam: 'view',
  style: {},
  dragZoomboxStyle: { backgroundColor: 'rgb(0, 50, 50, 0.5)' },
  translations: en
}

Map.propTypes = {
  /** callback passed by a <Provider> parent to attach props from this <Map> (map, selectInteraction, etc.) to context */
  addMapToContext: PropTypes.func,
  /** any ol-kit children components will automatically be passed a reference to the map object via the `map` prop */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** if this is set to false, the map will fill it's parent container */
  fullScreen: PropTypes.bool,
  /** place the ol-kit logo on the 'left', 'right', or set to 'none' to hide */
  logoPosition: PropTypes.string,
  /** optionally pass a custom map */
  map: PropTypes.object,
  /** callback called with initialized map object after optional animations complete
    note: if a Promise is returned from this function, Map will wait for onMapInit to resolve before rendering children
  */
  onMapInit: PropTypes.func,
  /** the length of debounce on map movements before the url gets updated */
  updateUrlDebounce: PropTypes.number,
  /** add map location coords + zoom level to url as query params */
  updateUrlFromView: PropTypes.bool,
  /** update map view based off the url param */
  updateViewFromUrl: PropTypes.bool,
  /** change the url param used to set the map location coords */
  urlViewParam: PropTypes.string,
  /** an openlayers select interaction passed down to connected components - created internally if not passed as prop */
  selectInteraction: PropTypes.object,
  /** apply inline styles to the map container */
  style: PropTypes.object,
  /** apply styles to the OL shift-zoom box */
  dragZoomboxStyle: PropTypes.object,
  /** object of string key/values (see: locales) */
  translations: PropTypes.object
}

export default connectToContext(Map)
