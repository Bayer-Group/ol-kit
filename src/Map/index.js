/*
  _Map is wrapped by a <Provider> to setup context
  without requiring a declartive <Provider> on the page
*/

export { default as Map } from './_Map'
export * from './utils'
