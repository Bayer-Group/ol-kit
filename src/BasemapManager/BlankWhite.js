import React from 'react'
import PropTypes from 'prop-types'
import olLayerVector from 'ol/layer/vector'
import olSourceVector from 'ol/source/vector'
import { BasemapOption, BasemapThumbnail, Label } from './styled'
import { connectToMap } from 'Map'

class BlankWhite extends React.Component {
  handleLayersChange = () => {
    this.forceUpdate()
  }

  componentDidMount () {
    this.props.map.getLayers().on('change', this.handleLayersChange)
  }

  componentWillUnmount () {
    this.props.map.getLayers().un('change', this.handleLayersChange)
  }

  onClick = () => {
    const layer = new olLayerVector({
      source: new olSourceVector()
    })
    const { map, layerTypeID, onBasemapChanged } = this.props

    layer[layerTypeID] = 'blankWhite' // make sure we can identify this layer as a layer that has been created from the ol-kit basemap component.

    const layers = map.getLayers()
    const layerArray = layers.getArray()
    const hasBasemap = layerTypeID && layerArray.length ? layerArray[0][layerTypeID] : false

    if (hasBasemap) {
      layers.setAt(0, layer)
    } else {
      layers.insertAt(0, layer)
    }
    layers.changed() // ol.Collection insertAt and setAt do not trigger a change event so we fire one manually so that we can rerender to display the active and inactive BasemapOptions
    onBasemapChanged(layer)
  }

  render () {
    const { translations, thumbnail, map, layerTypeID } = this.props
    const label = translations.label
    const layerArray = map.getLayers().getArray()
    const isActive = layerArray.length ? layerArray[0][layerTypeID] === 'blankWhite' : false

    return (
      <BasemapOption isActive={isActive} onClick={this.onClick}>
        <BasemapThumbnail thumbnail={thumbnail} />
        <Label>{label}</Label>
      </BasemapOption>
    )
  }
}

BlankWhite.propTypes = {
  map: PropTypes.object.isRequired,
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.shape({
    label: PropTypes.string
  }),
  /** A string containing an http url or data url to a thumbnail image */
  thumbnail: PropTypes.string,
  /** A unique string or symbol property name that will be set directly on the layer when it is created with a value containing a string identifying the type of basemap layer (e.g. '_ol_kit_basemap': 'osm').  This property should be a shared ID used to identify individual layers as 'basemap' layers.  */
  layerTypeID: PropTypes.oneOfType([PropTypes.symbol, PropTypes.string]),
  /** A callback that is fired when the basemap layer has been changed.  It is called with the updated layer. */
  onBasemapChanged: PropTypes.func
}

BlankWhite.defaultProps = {
  thumbnail: '',
  onBasemapChanged: () => {},
  translations: {
    label: 'Blank White'
  },
  layerTypeID: '_ol_kit_basemap'
}

export default connectToMap(BlankWhite)
