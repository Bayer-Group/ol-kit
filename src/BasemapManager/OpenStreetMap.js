import React from 'react'
import PropTypes from 'prop-types'
import olLayerTile from 'ol/layer/tile'
import olSourceOSM from 'ol/source/osm'
import { BasemapOption, BasemapThumbnail, Label } from './styled'
import { osm } from './thumbnails'
import { connectToMap } from 'Map'

class OpenStreetMap extends React.Component {
  handleLayersChange = () => {
    this.forceUpdate()
  }

  componentDidMount () {
    this.props.map.getLayers().on('change', this.handleLayersChange)
  }

  componentWillUnmount () {
    this.props.map.getLayers().un('change', this.handleLayersChange)
  }

  onClick = () => {
    const { map, layerTypeID, onBasemapChanged } = this.props
    const source = new olSourceOSM({
      layer: 'osm',
      cacheSize: 4096,
      reprojectionErrorThreshold: 0,
      crossOrigin: 'Anonymous'
    })
    const layer = new olLayerTile({
      preload: Infinity,
      extent: undefined,
      source
    })

    layer[layerTypeID] = 'osm' // make sure we can identify this layer as a layer that has been created from the ol-kit basemap component.

    const layers = map.getLayers()
    const layerArray = layers.getArray()
    const hasBasemap = layerTypeID && layerArray.length ? layerArray[0][layerTypeID] : false

    if (hasBasemap) {
      layers.setAt(0, layer)
    } else {
      layers.insertAt(0, layer)
    }
    layers.changed() // ol.Collection insertAt and setAt do not trigger a change event so we fire one manually so that we can rerender to display the active and inactive BasemapOptions
    onBasemapChanged(layer)
  }

  render () {
    const { translations, thumbnail, map, layerTypeID } = this.props
    const label = translations.label
    const layerArray = map.getLayers().getArray()
    const isActive = layerArray.length ? layerArray[0][layerTypeID] === 'osm' : false

    return (
      <BasemapOption isActive={isActive} onClick={this.onClick}>
        <BasemapThumbnail thumbnail={thumbnail} />
        <Label>{label}</Label>
      </BasemapOption>
    )
  }
}

OpenStreetMap.propTypes = {
  map: PropTypes.object.isRequired,
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.shape({
    label: PropTypes.string
  }),
  /** A string containing an http url or data url to a thumbnail image */
  thumbnail: PropTypes.string,
  /** A unique string or symbol property name that will be set directly on the layer when it is created with a value containing a string identifying the type of basemap layer (e.g. '_ol_kit_basemap': 'osm').  This property should be a shared ID used to identify individual layers as 'basemap' layers. */
  layerTypeID: PropTypes.oneOfType([PropTypes.symbol, PropTypes.string]),
  /** A callback that is fired when the basemap layer has been changed.  It is called with the updated layer. */
  onBasemapChanged: PropTypes.func
}

OpenStreetMap.defaultProps = {
  thumbnail: osm,
  onBasemapChanged: () => {},
  translations: {
    label: 'Open Street Map'
  },
  layerTypeID: '_ol_kit_basemap'
}

export default connectToMap(OpenStreetMap)
