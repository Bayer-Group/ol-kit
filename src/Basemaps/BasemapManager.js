import React from 'react'
import PropTypes from 'prop-types'
import { Container } from './styled'
import OpenStreetMap from './OpenStreetMap'
import BlankWhite from './BlankWhite'
import StamenTerrain from './StamenTerrain'
import StamenTonerDark from './StamenTonerDark'
import StamenTonerLite from './StamenTonerLite'

const LAYER_TYPE_ID = '_ol_kit_basemap'

/**
 * An easy Basemap switching component with customizable basemap options
 * @component
 * @category Basemap
 * @since 0.1.0
 */
class BasemapManager extends React.Component {
  static getLayerTypeID () {
    return LAYER_TYPE_ID
  }

  render () {
    const { children } = this.props

    return (
      <Container>
        <OpenStreetMap layerTypeID={LAYER_TYPE_ID} />
        <StamenTerrain layerTypeID={LAYER_TYPE_ID} />
        <StamenTonerDark layerTypeID={LAYER_TYPE_ID} />
        <StamenTonerLite layerTypeID={LAYER_TYPE_ID} />
        <BlankWhite layerTypeID={LAYER_TYPE_ID} />
        {children}
      </Container>
    )
  }
}

BasemapManager.propTypes = {
  /** Pass components as children of BasemapManager component */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]),
  /** Object with key/value pairs for translated strings */
  translations: PropTypes.object
}

export default BasemapManager
