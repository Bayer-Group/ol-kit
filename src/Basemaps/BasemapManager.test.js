import React from 'react'
import { render, waitFor } from '@testing-library/react'
import { prettyDOM } from '@testing-library/dom'
import { Map } from 'Map'
import BasemapManager from './BasemapManager'

describe('<BasemapManager />', () => {
  it.skip('should render a basic basemap manager component', async () => {
    const { container } = render(<Map><BasemapManager inlineProp={true} /></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
  it.skip('should render a single child', async () => {
    const child = <div id='child comp'>child comp</div>
    const { container } = render(<Map><BasemapManager>{child}</BasemapManager></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
  it.skip('should render an array of children', async () => {
    const child1 = <div key={1} id='1'>child comp</div>
    const child2 = <div key={2} id='2'>child comp</div>
    const child3 = <div key={3} id='3'>child comp</div>
    const { container } = render(<Map><BasemapManager>{[child1, child2, child3]}</BasemapManager></Map>)

    // wait for async child render
    await waitFor(() => {}, { container })

    expect(prettyDOM(container)).toMatchSnapshot()
  })
})
