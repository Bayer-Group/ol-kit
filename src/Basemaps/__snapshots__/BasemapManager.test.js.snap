// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasemapManager /> should render a basic basemap manager component 1`] = `
"[36m<div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhCb gdoKRR\\"[39m
    [33mid[39m=[32m\\"_ol_kit_map_test_id\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxjAm ehgTyw\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"sc-AxiKw cXKhWR\\"[39m
      [36m>[39m
        [0mPowered by ol-kit[0m
      [36m</p>[39m
      [36m<a[39m
        [33mclass[39m=[32m\\"sc-AxirZ flfhYH\\"[39m
        [33mhref[39m=[32m\\"https://ol-kit.com/\\"[39m
        [33mtarget[39m=[32m\\"_blank\\"[39m
        [33mtitle[39m=[32m\\"Powered by ol-kit\\"[39m
      [36m>[39m
        [36m<svg[39m
          [33mviewBox[39m=[32m\\"0 0 204.76 236.44\\"[39m
          [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
        [36m>[39m
          [36m<defs>[39m
            [36m<style>[39m
              [0m.cls-1[0m
              [0m{[0m
              [0mfill:#231f20[0m
              [0m}[0m
            [36m</style>[39m
          [36m</defs>[39m
          [36m<g[39m
            [33mid[39m=[32m\\"Layer_2\\"[39m
          [36m>[39m
            [36m<path[39m
              [33mclass[39m=[32m\\"cls-1\\"[39m
              [33md[39m=[32m\\"M102.38 0L0 59.11v118.22l102.38 59.11 102.38-59.11V59.11zM106 15.64l87 50.26v102L80.49 104.09zM18 62.27l74.66-43.1-33.95 118L18 62.27zM43.8 189l-32-18.51V75.37l42.35 77.86zm10.49 6.06l7.82-27.17 25.12 46.19zm50.38 26.44l-37.93-69.67 10.4-36.13 106.4 60.3z\\"[39m
              [33mid[39m=[32m\\"logo_examples\\"[39m
            [36m/>[39m
          [36m</g>[39m
        [36m</svg>[39m
      [36m</a>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"ol-viewport\\"[39m
      [33mstyle[39m=[32m\\"position: relative; overflow: hidden; width: 100%; height: 100%;\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-unselectable ol-layers\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; width: 100%; height: 100%; z-index: 0;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer-stopevent\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhUy jBsLOZ\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr hexoCE\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mOpen Street Map[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr gGrEoc\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Terrain[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr iuRyxP\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Dark[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr ioAVJR\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Lite[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr bCPvyl\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mBlank White[0m
      [36m</label>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</div>[39m"
`;

exports[`<BasemapManager /> should render a single child 1`] = `
"[36m<div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhCb gdoKRR\\"[39m
    [33mid[39m=[32m\\"_ol_kit_map_test_id\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxjAm ehgTyw\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"sc-AxiKw cXKhWR\\"[39m
      [36m>[39m
        [0mPowered by ol-kit[0m
      [36m</p>[39m
      [36m<a[39m
        [33mclass[39m=[32m\\"sc-AxirZ flfhYH\\"[39m
        [33mhref[39m=[32m\\"https://ol-kit.com/\\"[39m
        [33mtarget[39m=[32m\\"_blank\\"[39m
        [33mtitle[39m=[32m\\"Powered by ol-kit\\"[39m
      [36m>[39m
        [36m<svg[39m
          [33mviewBox[39m=[32m\\"0 0 204.76 236.44\\"[39m
          [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
        [36m>[39m
          [36m<defs>[39m
            [36m<style>[39m
              [0m.cls-1[0m
              [0m{[0m
              [0mfill:#231f20[0m
              [0m}[0m
            [36m</style>[39m
          [36m</defs>[39m
          [36m<g[39m
            [33mid[39m=[32m\\"Layer_2\\"[39m
          [36m>[39m
            [36m<path[39m
              [33mclass[39m=[32m\\"cls-1\\"[39m
              [33md[39m=[32m\\"M102.38 0L0 59.11v118.22l102.38 59.11 102.38-59.11V59.11zM106 15.64l87 50.26v102L80.49 104.09zM18 62.27l74.66-43.1-33.95 118L18 62.27zM43.8 189l-32-18.51V75.37l42.35 77.86zm10.49 6.06l7.82-27.17 25.12 46.19zm50.38 26.44l-37.93-69.67 10.4-36.13 106.4 60.3z\\"[39m
              [33mid[39m=[32m\\"logo_examples\\"[39m
            [36m/>[39m
          [36m</g>[39m
        [36m</svg>[39m
      [36m</a>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"ol-viewport\\"[39m
      [33mstyle[39m=[32m\\"position: relative; overflow: hidden; width: 100%; height: 100%;\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-unselectable ol-layers\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; width: 100%; height: 100%; z-index: 0; display: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer-stopevent\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhUy jBsLOZ\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr hexoCE\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mOpen Street Map[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr gGrEoc\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Terrain[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr iuRyxP\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Dark[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr ioAVJR\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Lite[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr bCPvyl\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mBlank White[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mid[39m=[32m\\"child comp\\"[39m
    [36m>[39m
      [0mchild comp[0m
    [36m</div>[39m
  [36m</div>[39m
[36m</div>[39m"
`;

exports[`<BasemapManager /> should render an array of children 1`] = `
"[36m<div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhCb gdoKRR\\"[39m
    [33mid[39m=[32m\\"_ol_kit_map_test_id\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxjAm ehgTyw\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"sc-AxiKw cXKhWR\\"[39m
      [36m>[39m
        [0mPowered by ol-kit[0m
      [36m</p>[39m
      [36m<a[39m
        [33mclass[39m=[32m\\"sc-AxirZ flfhYH\\"[39m
        [33mhref[39m=[32m\\"https://ol-kit.com/\\"[39m
        [33mtarget[39m=[32m\\"_blank\\"[39m
        [33mtitle[39m=[32m\\"Powered by ol-kit\\"[39m
      [36m>[39m
        [36m<svg[39m
          [33mviewBox[39m=[32m\\"0 0 204.76 236.44\\"[39m
          [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
        [36m>[39m
          [36m<defs>[39m
            [36m<style>[39m
              [0m.cls-1[0m
              [0m{[0m
              [0mfill:#231f20[0m
              [0m}[0m
            [36m</style>[39m
          [36m</defs>[39m
          [36m<g[39m
            [33mid[39m=[32m\\"Layer_2\\"[39m
          [36m>[39m
            [36m<path[39m
              [33mclass[39m=[32m\\"cls-1\\"[39m
              [33md[39m=[32m\\"M102.38 0L0 59.11v118.22l102.38 59.11 102.38-59.11V59.11zM106 15.64l87 50.26v102L80.49 104.09zM18 62.27l74.66-43.1-33.95 118L18 62.27zM43.8 189l-32-18.51V75.37l42.35 77.86zm10.49 6.06l7.82-27.17 25.12 46.19zm50.38 26.44l-37.93-69.67 10.4-36.13 106.4 60.3z\\"[39m
              [33mid[39m=[32m\\"logo_examples\\"[39m
            [36m/>[39m
          [36m</g>[39m
        [36m</svg>[39m
      [36m</a>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"ol-viewport\\"[39m
      [33mstyle[39m=[32m\\"position: relative; overflow: hidden; width: 100%; height: 100%;\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-unselectable ol-layers\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; width: 100%; height: 100%; z-index: 0; display: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"ol-overlaycontainer-stopevent\\"[39m
        [33mstyle[39m=[32m\\"position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;\\"[39m
      [36m/>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-AxhUy jBsLOZ\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr hexoCE\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mOpen Street Map[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr gGrEoc\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Terrain[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr iuRyxP\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Dark[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr ioAVJR\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mStamen Toner Lite[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"sc-AxheI bIMtVZ _ol_kit_basemapOption\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-Axmtr bCPvyl\\"[39m
      [36m/>[39m
      [36m<label[39m
        [33mclass[39m=[32m\\"sc-AxmLO cPQRqo\\"[39m
      [36m>[39m
        [0mBlank White[0m
      [36m</label>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mid[39m=[32m\\"1\\"[39m
    [36m>[39m
      [0mchild comp[0m
    [36m</div>[39m
    [36m<div[39m
      [33mid[39m=[32m\\"2\\"[39m
    [36m>[39m
      [0mchild comp[0m
    [36m</div>[39m
    [36m<div[39m
      [33mid[39m=[32m\\"3\\"[39m
    [36m>[39m
      [0mchild comp[0m
    [36m</div>[39m
  [36m</div>[39m
[36m</div>[39m"
`;
